File: localization/pf/python/pf.py
import numpy as np

class ParticleFilter:
    def __init__(self, num_particles):
        self.num_particles = num_particles
        self.particles = np.zeros((num_particles, 3))  # [x, y, theta]
        self.weights = np.ones(num_particles) / num_particles

    def init(self, x_range, y_range, theta_range):
        self.particles[:, 0] = np.random.uniform(x_range[0], x_range[1], self.num_particles)
        self.particles[:, 1] = np.random.uniform(y_range[0], y_range[1], self.num_particles)
        self.particles[:, 2] = np.random.uniform(theta_range[0], theta_range[1], self.num_particles)

    def predict(self, v, w, dt):
        theta = self.particles[:, 2]
        self.particles[:, 0] += v * dt * np.cos(theta)
        self.particles[:, 1] += v * dt * np.sin(theta)
        self.particles[:, 2] += w * dt

    def update(self, z, std):
        z_x, z_y = z
        distances = np.linalg.norm(self.particles[:, :2] - np.array([z_x, z_y]), axis=1)
        self.weights = np.exp(-(distances ** 2) / (2 * std[0] ** 2))
        self.weights += 1.e-300  # avoid division by zero
        self.weights /= self.weights.sum()

    def resample(self):
        indices = np.arange(self.num_particles)
        cumulative_sum = np.cumsum(self.weights)
        cumulative_sum[-1] = 1.0  # avoid round-off error
        random_values = np.random.rand(self.num_particles)
        new_indices = np.searchsorted(cumulative_sum, random_values)
        self.particles[:] = self.particles[new_indices]
        self.weights.fill(1.0 / self.num_particles)

if __name__ == "__main__":
    pf = ParticleFilter(num_particles=100)
    pf.init(x_range=(0, 1), y_range=(0, 1), theta_range=(0, 2 * np.pi))

    v = 1.0  # velocity
    w = 0.1  # angular velocity
    dt = 0.1  # time step

    pf.predict(v, w, dt)

    z = np.array([1.0, 0.1])
    std = [0.1, 0.1]

    pf.update(z, std)
    pf.resample()

    for i, particle in enumerate(pf.particles):
        print(f"Particle {i}: x={particle[0]:.2f}, y={particle[1]:.2f}, theta={particle[2]:.2f}, weight={pf.weights[i]:.2f}")

File: localization/pf/c/pf.c
#include "pf.h"

void pf_init(ParticleFilter *pf, int num_particles) {
    pf->num_particles = num_particles;
    pf->particles = (Particle *)malloc(num_particles * sizeof(Particle));
    for (int i = 0; i < num_particles; ++i) {
        pf->particles[i].x = rand() / (double)RAND_MAX;
        pf->particles[i].y = rand() / (double)RAND_MAX;
        pf->particles[i].theta = rand() / (double)RAND_MAX * 2 * M_PI;
        pf->particles[i].weight = 1.0 / num_particles;
    }
}

void pf_predict(ParticleFilter *pf, double v, double w, double dt) {
    for (int i = 0; i < pf->num_particles; ++i) {
        double theta = pf->particles[i].theta;
        pf->particles[i].x += v * dt * cos(theta);
        pf->particles[i].y += v * dt * sin(theta);
        pf->particles[i].theta += w * dt;
    }
}

void pf_update(ParticleFilter *pf, double z_x, double z_y, double std[]) {
    double sum_weights = 0.0;
    for (int i = 0; i < pf->num_particles; ++i) {
        double dx = z_x - pf->particles[i].x;
        double dy = z_y - pf->particles[i].y;
        double distance = sqrt(dx * dx + dy * dy);
        double weight = exp(-(distance * distance) / (2 * std[0] * std[0]));
        pf->particles[i].weight = weight;
        sum_weights += weight;
    }
    for (int i = 0; i < pf->num_particles; ++i) {
        pf->particles[i].weight /= sum_weights;
    }
}

void pf_resample(ParticleFilter *pf) {
    Particle *new_particles = (Particle *)malloc(pf->num_particles * sizeof(Particle));
    double beta = 0.0;
    int index = rand() % pf->num_particles;
    double max_weight = 0.0;
    for (int i = 0; i < pf->num_particles; ++i) {
        if (pf->particles[i].weight > max_weight) {
            max_weight = pf->particles[i].weight;
        }
    }
    for (int i = 0; i < pf->num_particles; ++i) {
        beta += ((rand() / (double)RAND_MAX) * 2.0 * max_weight);
        while (beta > pf->particles[index].weight) {
            beta -= pf->particles[index].weight;
            index = (index + 1) % pf->num_particles;
        }
        new_particles[i] = pf->particles[index];
    }
    free(pf->particles);
    pf->particles = new_particles;
}

void pf_free(ParticleFilter *pf) {
    free(pf->particles);
}
